Better presets

Wueffi opened the pull request:
This adds the birdflop presets and some player presets to the plugin and organizes them in a different (better) way.

/nick presets <pride|birdflop|players>

Wueffi committed:
Create birdflop-presets.kt

Wueffi committed:
Add all presets

Wueffi committed:
Update ChattORESpec.kt

Wueffi committed:
Update ChattORESpec.kt

Wueffi committed:
Update birdflop-presets.kt

Wueffi committed:
Create player-presets.kt

Wueffi committed:
Rename Pride.kt to pride-presets.kt

Wueffi committed:
Update Nick.kt

Wueffi committed:
Update README.md

Wueffi committed:
Update README.md

Wueffi committed:
Update README.md

Wueffi committed:
Update README.md

Nickster258 commented on src/main/kotlin/birdflop-presets.kt:
```diff
@@ -0,0 +1,50 @@
+
+val birdflopGradients = mapOf(
+    "birdflop" to arrayOf(
+        "##084CFB",
```
Hex color code formatting.

Nickster258 commented on src/main/kotlin/birdflop-presets.kt:
```diff
@@ -0,0 +1,50 @@
+    ),
+    "Mango" to arrrayOF (
+        "#1488CC",
+        "2B32B2"
```
Hex color code formatting.

Nickster258 commented on src/main/kotlin/commands/Nick.kt:
```diff
@@ -89,10 +89,17 @@ class Nick(private val chattORE: ChattORE) : BaseCommand() {
+        
         val renderedPresets = ArrayList<Component>()
-        for ((presetName, preset) in chattORE.config[ChattORESpec.nicknamePresets]) {
+        for ((presetName, preset) in presets) {
```
Because of this modification, the config file will never be read when loading. This won't allow us to modify _just_ the config to change what presets are available.

Wueffi replied:
Yea didnt think abt that

Nickster258 commented on src/main/kotlin/commands/Nick.kt:
```diff
@@ -89,10 +89,17 @@ class Nick(private val chattORE: ChattORE) : BaseCommand() {
+    @CommandCompletion("pride|birdflop|players")
+    fun presets(player: Player, @Optional shownText: String?, @Optional type: String?) {
+        val presets = when (type?.lowercase()) {
+            "pride" -> pridePresets
```
It is probably better to separate each of these types of presets into separate configuration items. It may even be better to make the presets a map of keys (preset category) associated with a map itself of keys (preset name) and the preset value itself. This would allow us to define more types of presets later.

Wueffi replied:
Yea I like the map of maps system with the nickname_presets [all presets combined] > categorys [e. g. pride] > preset [e.g. nickster]

Nickster258 commented on src/main/kotlin/birdflop-presets.kt:
```diff
@@ -0,0 +1,50 @@
+    ),        
+)
+
+val birdflopPresets = birdflopGradients.mapValues { (_, colors) ->
```
This is a little duplicated from the `pridePresets.kt`.

Wueffi replied:
Is that bad?

Nickster258 replied:
I'd prefer to not have the same code repeated in 3 places. Since this is only really used in the ChattORESpec.kt, it could be added there as an extension function, potentially defined like this:

```kotlin
fun Map<String, Array<String>>.toGradientMinimessage() = this.mapValues { (_, colors) ->
    "<gradient:${colors.joinToString(':'.toString())}><username></gradient>"
}.toSortedMap()
```

Nickster258 commented on src/main/kotlin/player-presets.kt:
```diff
@@ -0,0 +1,13 @@
+    ),
+)
+
+val playerPresets = playerColors.mapValues { (_, colors) ->
```
This is a little duplicated from the `pridePresets.kt`.

Nickster258 replied:
Same comment on the extension function.

Nickster258 commented on src/main/kotlin/entity/ChattORESpec.kt:
```diff
@@ -46,5 +48,8 @@ object ChattORESpec : ConfigSpec("") {
     }
 
-    val nicknamePresets by optional(pridePresets)
+    val nicknamePresetsPride by optional(pridePresets)
```
My thought on the map of maps idea for organizing presets could make the resulting YAML turn into something like this:

```yaml
nickname_presets:
    pride:
        one: 'preset'
        two: 'preset'
    birdflop:
        ....
    players:
        ....
    anotherType:
        one: 'preset'
        two: 'preset'
````

Wueffi replied:
What abt turning all 3 example preset files into 1 big presets.kt file with this structure

Nickster258 reviewed (changes requested):
Overall I like the idea of categorizing types of presets separately. It makes seeing the available presets much less cluttered.

I added comments regarding this specific implementation and how it can be easily modified for future categories. Also a comment on how this implementation ignores the config file when loading, which would have to be fixed so that we can make changes in the config that get loaded by the proxy.

Wueffi committed:
Update birdflop-presets.kt

Wueffi committed:
Update birdflop-presets.kt

Wueffi committed:
Update Nick.kt

Wueffi committed:
Different structure

Wueffi committed:
Update Config

Wueffi committed:
Added support for config

Wueffi commented:
@Nickster258 do you wanna take a look at it? I updated the chattorespec and nick.kt to fit the yaml config structure

Wueffi commented:
Also i accidently clicked on review nickstore this morning 0.o but you can take a look at it now :]

Nickster258 commented on src/main/kotlin/entity/ChattORESpec.kt:
```diff
@@ -45,6 +47,10 @@ object ChattORESpec : ConfigSpec("") {
         val joinDiscord by optional("**<player> has joined the network**")
         val leaveDiscord by optional("**<player> has left the network**")
     }
+    object nicknamePresets : ConfigSpec() {
```
I was thinking more like this:
```kotlin
val nicknamePresets by optional(
    mapOf(
        "pride" to pridePresets,
        "birdflop" to birdflopPresets,
        "players" to playerPresets
    )
)
```

This would allow us to create entirely new groups without modifying the Konf configuration.

Nickster258 commented on src/main/kotlin/birdflop-presets.kt:
```diff
@@ -0,0 +1,50 @@
+        "#084CFB",
+        "#ADF3FD"
+    ),
+    "SimplyMC" to arrrayOF (
```
I just realized this is calling `arrayOF` but it should be `arrayOf`. All subsequent calls in this file are also `arrayOF`.

Nickster258 commented on src/main/kotlin/player-presets.kt:
```diff
@@ -0,0 +1,13 @@
+
+package chattore
+val playerColors = mapOf(
+    "Waffle" to arrrayOF (
```
Same comment on `arrayOF` needing to be `arrayOf`.

Nickster258 commented on src/main/kotlin/commands/Nick.kt:
```diff
@@ -89,10 +89,19 @@ class Nick(private val chattORE: ChattORE) : BaseCommand() {
+    @CommandCompletion("pride|birdflop|players")
+      fun presets(player: Player, @Optional shownText: String?, @Optional type: String?) {
+        val additionalPresets = when (type?.lowercase()) {
+            "pride" -> chattORE.config[ChattORESpec.nicknamePresets.pride]
```
My comment on the config file specification change would change this.

Wueffi committed:
Fixed arrayOf

Wueffi committed:
Fixed arrayOf

Wueffi committed:
Cleaned up the mess and the duplicated things

Wueffi committed:
Changed nicknamePresets in config

Wueffi committed:
Fixed Nick.kt preset loading

Wueffi commented:
I think I added all req. changes while keeping the structure you wanted :]

btw I combined the 3 preset files into 1 file but kept the togradient function in the config file.

Wueffi committed:
Update presets.kt 🇩🇪

how did I write arrrayOF i wasnt even drunk or smth

Yes ik beautiful commit message

Nickster258 commented on src/main/kotlin/presets.kt:
The file name should be `Presets.kt`, capitalized.

Nickster258 commented on src/main/kotlin/entity/ChattORESpec.kt:
```diff
@@ -45,6 +47,12 @@ object ChattORESpec : ConfigSpec("") {
     }
+    val nicknamePresets by optional(
+        mapOf(
+            "pride" to prideColors.toGradientMinimessage(),
```
Where is `toGradientMinimessage()` defined? I don't see it anywhere.

Wueffi replied:
declared a function in config seems like i added it with my 2n last commit but deleted it again with my last commit 2 months ago lmao

Nickster258 commented on src/main/kotlin/entity/ChattORESpec.kt:
```diff
@@ -2,6 +2,8 @@ package chattore.entity
 
 import com.uchuhimo.konf.ConfigSpec
 import chattore.pridePresets
```
These are referred to as `presets` here but `colors` later.

Wueffi replied:
yea I took that part from the presets.kt for the config

val pridePresets = prideColors.toGradientMinimessage()
val birdflopPresets = birdflopColors.toGradientMinimessage()
val playerPresets = playerColors.toGradientMinimessage()

i did prideColors instead of pridePresets now

Nickster258 reviewed (changes requested):
It would help tremendously to review this if you were testing / compiling them yourself. As it currently stands, this PR doesn't build.

Wueffi committed:
spel fix

Wueffi committed:
fix presets then colors

Wueffi committed:
decalred a function in config :skull:

Wueffi commented:
code ded
